src/psum.c

```diff
@@ -0,0 +1,184 @@
+#include "data.table.h"
+
+// like base::p{max,min}, but for sum
+SEXP psum(SEXP x, SEXP narmArg) {
+  if (!isNewList(x)) error(_("Internal error: x must be a list"));
+  if (!isLogical(narmArg) || LENGTH(narmArg)!=1 || LOGICAL(narmArg)[0]==NA_LOGICAL) error(_("na.rm must be TRUE or FALSE"));
+
+  int J=LENGTH(x);
+  SEXPTYPE outtype = INTSXP;
+  int n = -1;
+  for (int j=0; j<J; j++) {
+    SEXP xj=VECTOR_ELT(x, j);
+    switch(TYPEOF(xj)) {
+    case INTSXP:
+      break;
+    case REALSXP:
+      outtype = REALSXP;
+      break;
+    default:
+      error(_("Only numeric inputs are supported for psum"));
+    }
+    if (n >= 0) {
+      if (n != LENGTH(xj)) {
+        error(_("Inconsistent input lengths -- first found %d, but %d element has length %d"), n, j+1, LENGTH(xj));
+      }
+    } else { // initialize
+      n = LENGTH(xj);
+    }
+  }
+
+
+  // TODO: support int->double conversion
+  SEXP out = PROTECT(allocVector(outtype, n));
+  if (LOGICAL(narmArg)[0]) {
+    if (outtype == INTSXP) {
+      int *outp = INTEGER(out);
+      // initialize to NA to facilitate all-NA rows --> NA output
+      for (int i=0; i<n; i++) outp[i] = NA_INTEGER;
```

Comment:
I recall `allocNAvectorLike` somewhere, can't we use it here?

Reply:
used `writeNA` directly since we have `out` declared & we only want to fill NA in the `na.rm=TRUE` branch